src/connector.rs

```diff
@@ -0,0 +1,582 @@
+    }
+
+    /// FT transfer call from sender account (invoker account) to receiver
+    pub fn ft_transfer_call(&mut self) {
+        sdk::assert_one_yocto();
+        let args =
+            TransferCallCallArgs::try_from_slice(&sdk::read_input()).expect(ERR_FAILED_PARSE);
+        #[cfg(feature = "log")]
+        sdk::log(&format!(
+            "Transfer call to {} amount {}",
+            args.receiver_id, args.amount,
+        ));
+
+        self.ft
+            .ft_transfer_call(&args.receiver_id, args.amount, &args.memo, args.msg);
+    }
+
+    /// FT storage deposit logic
+    pub fn storage_deposit(&mut self) {
+        let args =
+            StorageDepositCallArgs::try_from_slice(&sdk::read_input()).expect(ERR_FAILED_PARSE);
+        let res = self
+            .ft
+            .storage_deposit(args.account_id.as_ref(), args.registration_only)
+            .try_to_vec()
+            .unwrap();
+        self.save_contract();
+        sdk::return_output(&res[..]);
+    }
+
+    /// FT storage withdraw
+    pub fn storage_withdraw(&mut self) {
+        sdk::assert_one_yocto();
+        let args =
+            StorageWithdrawCallArgs::try_from_slice(&sdk::read_input()).expect(ERR_FAILED_PARSE);
+        let res = self.ft.storage_withdraw(args.amount).try_to_vec().unwrap();
+        self.save_contract();
+        sdk::return_output(&res[..]);
+    }
+
+    /// Get balance of storage
+    pub fn storage_balance_of(&self) {
+        let args =
+            StorageBalanceOfCallArgs::try_from_slice(&sdk::read_input()).expect(ERR_FAILED_PARSE);
+        let res = self
+            .ft
+            .storage_balance_of(&args.account_id)
+            .try_to_vec()
+            .unwrap();
+        sdk::return_output(&res[..]);
+    }
+
+    /// ft_on_transfer callback function
+    pub fn ft_on_transfer(&mut self, engine: &Engine) {
+        #[cfg(feature = "log")]
+        sdk::log("Call ft_on_trasfer");
+        let args = FtOnTransfer::try_from_slice(&sdk::read_input()).expect(ERR_FAILED_PARSE);
+        let predecessor_account_id = String::from_utf8(sdk::predecessor_account_id()).unwrap();
+        let current_account_id = String::from_utf8(sdk::current_account_id()).unwrap();
+        // Parse message with specific rules
+        let message_data = self.parse_on_transfer_message(&args.msg);
+
+        // Special case when predecessor_account_id is current_account_id
+        if current_account_id == predecessor_account_id {
+            let fee = message_data.fee.as_u128();
+            self.burn_near(current_account_id, args.amount);
+            self.mint_eth(message_data.recipient, args.amount - fee);
+            // Mint fee to relayer
+            if fee > 0 {
+                let evm_relayer_address: EthAddress = engine
+                    .get_relayer(&message_data.relayer.as_bytes())
+                    .expect("ERR_WRONG_RELAYER_ID")
+                    .0;
+                self.mint_eth(evm_relayer_address, fee);
+            }
```

Comment:
This part works incorrectly and actually is critical. Currently, this works like that:

1) If the fee is zero, the user receives the full amount, and the transaction is "Succeeded".
2) If the fee is greater than thero:
    a) Query for the registered relayer, the relayer is registered. Then mint appropriate amount of fee to the relayer. The transaction is "Succeeded".
    b) Query for the registered relayer, the relayer is NOT registered. So we panic with "ERR_WRONG_RELAYER_ID", the transaction is "Failed". The user is not receiving his money, the proof was recorded as used on the previous steps so it's not possible to finish the deposit for this transaction any more. So basically, it's easy to burn any produced proof by not registering the relayer.

The suggested solution for that is the following:
* Make sure we never ever panic in the `ft_on_transfer` execution (at least for the ETH branch), as the proof was recorded on the previous steps so we won't be able to restore it.
* If the `fee > 0`, but there's no registered relayer for the predecessor account, mint the fee to the recipient as well. If we don't mint the fee for the recipient, this amount won't be minted and there's no sense in it.

Reply:
@mrLSD @joshuajbouw @artob

Reply:
Note also the todo

Reply:
Hugely important on what you said. If there is a proof, we need to be able to use it on the previous steps. They must ALWAYS be restorable, only once, of course.

Reply:
There's no way to straight-forwardly revert the chain of promises at the moment. So this is correct that the proof is recorded in the previous step and only after that the tokens are minted. This protects from proof reusage and possible reentrancy attempts. We just need to try making all possible checks before storing the proof and don't let it panic at a later stage.

Reply:
Good point!
@joshuajbouw FYI it's not related to Proof. We can reproduce it without `Deposit` call because ft_transfer_call is a public method.

Reply:
@sept-en @jakestutzman I fixed that issue. Also, I placed an early checker for all possible assertions in ft_on_transfer.